pytrickle/client.py

```diff
@@ -108,6 +112,18 @@ async def start(self, request_id: str = "default"):
         finally:
             self.running = False
             logger.info("Stopping protocol due to client loops ending")
+            
+            # Call the optional on_stream_stop callback before stopping protocol
+            if hasattr(self.frame_processor, 'on_stream_stop') and self.frame_processor.on_stream_stop:
```

Comment:
Do we need the `hasattr`? Using this change will always have the attribute right?

Should not be a performance issue though since this is on startup.

Reply:
No, it isn't an abstract method of frame_processor. It's appended by StreamProcessor on initialization like other callbacks.

`StreamProcessor` accepts an `on_stream_stop` callback as a parameter, similar to `param_updater`. `_InternalFrameProcessor` extends the abstract `FrameProcessor` class and stores the `on_stream_stop` callback as an attribute

Reply:
In this case, on_stream_stop is more of an event triggered by the client, rather than a method called by the FrameProcessor. I added it as an abstract class now so it is available to implement in frame processors https://github.com/livepeer/pytrickle/pull/25/commits/f347e41fdc47a04841159162b10d00765a7be7bc

 Client still needs to call `on_stream_stop` at this point in the protocol shutdown sequence. There is no other coordination with FrameProcessors currently afaik. https://github.com/livepeer/pytrickle/blob/f347e41fdc47a04841159162b10d00765a7be7bc/pytrickle/client.py#L117-L124

Reply:
> It seems strange to call something from TrickleClient that is setup in a higher level abstraction.
Yeah, it sounds off because `TrickleClient` was originally intended as a class for interacting with trickle protocol directly in a multimedia context. In a consumer context it's an internal component so the term `client` is a bit misleading. 

I think once we reorganize classes into package namespaces this will be more clear